alf/algorithms/mbrl_algorithm.py

```diff
@@ -0,0 +1,172 @@
+
+        """
+        train_state_spec = MbrlState(
+            dynamics=dynamics_module.train_state_spec, reward=(), planner=())
+
+        super().__init__(
+            feature_spec,
+            action_spec,
+            train_state_spec=train_state_spec,
+            gradient_clipping=gradient_clipping,
+            debug_summaries=debug_summaries,
+            name=name)
+
+        flat_action_spec = tf.nest.flatten(action_spec)
+        action_spec = flat_action_spec[0]
+
+        assert not tensor_spec.is_discrete(action_spec), "only support \
+                                                    continious control"
+
+        num_actions = action_spec.shape[-1]
+
+        flat_feature_spec = tf.nest.flatten(feature_spec)
+        assert len(flat_feature_spec) == 1, "Mbrl doesn't support nested \
+                                             feature_spec"
+
+        feature_dim = flat_feature_spec[0].shape[-1]
+
+        self._action_spec = action_spec
+        self._num_actions = num_actions
+
+        self._dynamics_module = dynamics_module
+        self._reward_module = reward_module
+        self._planner_module = planner_module
+        self._planner_module.set_reward_func(self._calc_step_reward)
+        self._planner_module.set_dynamics_func(self._predict_next_step)
+
+    def _predict_next_step(self, time_step, state):
+        """Predict the next step (observation and state) based on the current
+            time step and state
+        Args:
+            time_step (ActionTimeStep): input data for next step prediction
+            state (MbrlState): input state next step prediction
+        Returns:
+            next_time_step (ActionTimeStep): updated time_step with observation
+                predicted from the dynamics module
+            next_state (MbrlState): updated state from the dynamics module
+        """
+        dynamics_step = self._dynamics_module.predict(time_step,
+                                                      state.dynamics)
+        next_time_step = time_step._replace(observation=dynamics_step.outputs)
+        next_state = state._replace(dynamics=dynamics_step.state)
+        return next_time_step, next_state
+
+    def _calc_step_reward(self, obs, action):
+        reward = self._reward_module.compute_reward(obs, action)
+        return reward
+
+    def _predict_with_planning(self, time_step: ActionTimeStep, state):
+        action = self._planner_module.generate_plan(time_step, state)
+        dynamics_state = self._dynamics_module.update_state(
+            time_step, state.dynamics)
+
+        return PolicyStep(
+            action=action,
+            state=MbrlState(dynamics=dynamics_state, reward=(), planner=()),
+            info=MbrlInfo())
+
+    def predict(self, time_step: ActionTimeStep, state, epsilon_greedy=1.):
+        return self._predict_with_planning(time_step, state)
+
+    def rollout(self, time_step: ActionTimeStep, state, mode):
```

Comment:
rollout => rollout_step
I wonder how the your example can successfully run with incorrect function names of predict and rollout.

Reply:
Most of the function/data-structure names issues you noticed are because the TF-version (first commit) has been looked into.
I included it mainly for reference purpose. Sorry for the confusion.